### PR TITLE
Semantic index proposal2

### PR DESCRIPTION
**Pull request by javiertoledo**
I continued with the changes made by @adrian-lorenzo, trying to figure out a way to provide abstract components that need to be initialized with specific implementations. In this example. I've built a service that exposes the `learn` and `search` endpoints, configured to use the OpenAI embeddings model and Pinecone (stubbed for now).

This is what I've done so far:

1.  I've created two new abstract components: An embeddings model and a vector store, which are defined as follows:

```kotlin
data class EmbeddingsModel(
    @Mandatory val name: String,
    @Mandatory var embed: EmbedHandler
) : PrimitiveModule() {} // Here I'm not sure about the value of extending PrimitiveModule, so this is one of the things to discuss


data class VectorStore(
    @Mandatory val name: String,
    @Mandatory var store: StoreHandler,
    @Mandatory var search: SearchHandler
) : PrimitiveModule() {}
```

2. With these two abstract components, I can write a generic implementation of a semantic search:

```kotlin
class SemanticSearch(
    @Mandatory val name: String,
    @Mandatory var embeddingsModel: EmbeddingsModel,
    @Mandatory var vectorStore: VectorStore
) : PrimitiveModule() {
    var learn = handler("learn") { text: String ->
        val embedding = embeddingsModel.embed(text)
        val semanticEntry = SemanticEntry(text, embedding = embedding)
        vectorStore.store(semanticEntry)
        semanticEntry
    }

    var search = handler("learn") { text: String ->
        val embedding = embeddingsModel.embed(text)
        val semanticEntry = SemanticEntry(text, embedding = embedding)
        vectorStore.search(semanticEntry)
    }
}

fun semanticSearch(name: String, embeddingsModel: EmbeddingsModel, vectorStore: VectorStore): SemanticSearch =
    makeDsl(SemanticSearch(name, embeddingsModel, vectorStore)) {}
```

3. Then, I built a couple of builder functions for the openAI's and Pinecone APIs:

```kotlin
fun pineconeVectorStore(
    apiKey: String,
    indexName: String
): VectorStore {
    return VectorStore(
        "pinecone",
        handler("store") { entry: SemanticEntry ->
            // TODO: Implement Pinecone store
            {}
        },
        handler("search") { entry: SemanticEntry ->
            // TODO: Implement Pinecone search
            listOf(entry)
        }
    )
}

fun openAIEmbeddingsModel(apiKey: String): EmbeddingsModel {
    return EmbeddingsModel("openAI", handler("embed")
    { text: String ->
        // TODO: Initialize and call the actual OpenAI API
        arrayOf(
            0.0,
            0.0
        )
    })
}
```

3. With all these pieces, I can quickly build a semantic index service like this:

```kotlin
fun main() {
    api("SemanticSearchService") {
        with(
            semanticSearch(
                "semanticSearch",
                openAIEmbeddingsModel("someAPIKey"),
                pineconeVectorStore("someOtherAPIKey", "someVectorStoreName")
            )
        ) {
            write(learn)
            read(search)
        }
    }
        .toHttpApp()
        .start()
        .block()
}
```

I think that @NickSeagull showed me a way to do this with the `use` method instead of the `with`, and I'm still not totally familiar with this component DSL, so this is more a proposal for discussion than an actual implementation proposal.

One thing to notice is that this DSL helps to compose modules in a Kotlin application, but I think we should aim to have a plugins system where the configuration is made outside of Kotlin in a config file. As this DSL is essentially data, if we're able to load the JARs of the specific implementations automatically, we could design a generic API service that loads the config from a JSON. This will be helpful to automate the deployments if we integrate this with the SaaS selector.

### CODE REVIEW

File: modules/semantic-search/src/main/kotlin/com/theagilemonkeys/llmental/semanticsearch/SemanticSearch.kt [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+import com.theagilemonkeys.llmental.core.Mandatory
+import com.theagilemonkeys.llmental.core.PrimitiveModule
+import com.theagilemonkeys.llmental.core.handler.handler
+import com.theagilemonkeys.llmental.core.makeDsl
+import com.theagilemonkeys.llmental.core.schema.SemanticEntry
+import com.theagilemonkeys.llmental.embeddingsModel.EmbeddingsModel
+import com.theagilemonkeys.llmental.vectorStore.VectorStore
+
+data class SemanticSearch(
+    @Mandatory val name: String,
```
Comment (adrian-lorenzo) [inline modules/semantic-search/src/main/kotlin/com/theagilemonkeys/llmental/semanticsearch/SemanticSearch.kt:12]: What's the value of the `@Mandatory` annotation? @NickSeagull @javiertoledo 

As far as I know, Kotlin has optional values and by default, values can't be null

File: modules/semantic-search/src/main/kotlin/com/theagilemonkeys/llmental/semanticsearch/SemanticSearch.kt [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+    @Mandatory var embeddingsModel: EmbeddingsModel,
+    @Mandatory var vectorStore: VectorStore
+) : PrimitiveModule() {
+    var learn = handler("learn") { text: String ->
+        val embedding = embeddingsModel.embed(text)
+        val semanticEntry = SemanticEntry(text, embedding = embedding)
+        vectorStore.store(semanticEntry)
+        semanticEntry
+    }
```
Comment (adrian-lorenzo) [inline modules/semantic-search/src/main/kotlin/com/theagilemonkeys/llmental/semanticsearch/SemanticSearch.kt:21]: I guess this is not `learn` but `store`, right?

**Review by adrian-lorenzo — COMMENTED** [review]
I think this approach makes sense, but I have difficulty differentiating `handlers` and `PrimitiveModule` now. Also, `PrimitiveModule` seems to be extremely generic in its behaviour. It seems like it can do whatever someone pleases.

**Comment by javiertoledo** [comment]
Yes, I think we should reconsider the design, it feels a bit redundant, as I'm basically just using constructors. Let's discuss it a bit today.

**Comment by NickSeagull** [comment]
Closing as outdated

**Pull request closed without merging**